File: src/CheetahRoller.Domain/RollSpec.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using CheetahRoller.Base.Extensions;

//using CheetahRoller.Base;

namespace CheetahRoller.Domain
{
    public struct RollSpec
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="spec"></param>
        /// <exception cref="InvalidSpecException">When spec fails to match expected regex</exception>
        public RollSpec(string spec)
        {
            if (string.IsNullOrEmpty(spec))
            {
                throw new ArgumentException("Spec must not be empty", nameof(spec));
            }

            var matches = SpecRegex.Matches(spec).ToList();

            if (!matches.Any(m => m.Success))
            {
                throw new InvalidSpecException();
            }

            Dice = matches
                .Where(m => m.Success)
                .SelectMany(m => Enumerable.Range(0, int.Parse(m.Groups["count"].Value))
                    .Select(r => new DiceSize(int.Parse(m.Groups["size"].Value))));
        }

        // 4d6
        public static readonly Regex SpecRegex = new Regex(@"(?<count>\d+)d(?<size>\d+)", RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);

        public class InvalidSpecException : Exception
        {
        }

        public IEnumerable<DiceSize> Dice { get; set; }
    }
}

File: src/components/rollSpec.js
class RollSpecParser {
    constructor(spec){
        var specRegex = new RegExp('(\\d*)d(\\d+)', 'gi');

        var match = specRegex.exec(spec);
        this._specs = [];

        while (match != null){
            this._specs.push(new RollSpec(match[1], match[2]));
            match = specRegex.exec(spec);
        }
    }

    get specs() {return this._specs;}

    get isValid() {return this.specs.length > 0;}
}

export default RollSpecParser;

class RollSpec{
    constructor(count, size){
        this._count = count || 1;
        this._size = size;
    }

    get count() {return this._count;}

    get size() {return this._size;}
}